### PR TITLE
File reading function moved to new parent class & pathlib support

### PR DESCRIPTION
**Pull request by tetov**
`BaseReader.read_from_location` returns an iterable for lines in file.

| File format | ASCII | Binary | Skeleton? | BaseReader implemented    |
| ----------- | ----- | ------ | --------- | ------------------------- |
| `amf`       | xml   |        | x         | x                         |
| `dxf`       | x     | x      | x         | x                         |
| `las`       |       | x      | x         | x                         |
| `obj`       | x     |        |           | x                         |
| `off`       | x     |        |           | x                         |
| `ply`       | x     | x      |           | For ASCII                 |
| `stl`       | x     | x      |           | Not yet                   |
| `urdf`     | xml   |        |           | Not yet                   |
| `xml`       | x     |        |           | Needs whole file to parse?|
| `xml_cli`   | x     |        |           | Needs whole file to parse?|

### What type of change is this?

- [x] New feature in a **backwards-compatible** manner.

### Checklist

1. [x] Add the change to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading: `Added`, `Changed`, `Removed`.
1. [x] Run all tests on your computer (i.e. `invoke test`).
1. [x] If you add new functions/classes, check that:
   1. [x] Are available on a second-level import, e.g. `compas.datastructures.Mesh`.
   1. [x] Add unit tests (especially important for algorithm implementations).

### CODE REVIEW

**Comment by tetov** [comment]
I will look into adding the same functionality for binary files, either keep this PR open until that is done or merge this and I'll submit a new PR later on.

**Questions**
* Would you like me to move the tests in `tests/compas/files/test_stl.py` to `tests/compas/datastructures/test_mesh.py`? Could be fitting to do it in this PR.
* Would you like me to put the `BaseReader` class in a file with another name, e.g. `_utils.py`?
* I'm guessing that iterating over lines won't work for the XML reader, am I right in that assumption?

**Comment by tetov** [comment]
I updated questions in above comment. :)

**Review by Licini — APPROVED** [review]
LGTM

**Comment by Licini** [comment]
> I will look into adding the same functionality for binary files, either keep this PR open until that is done or merge this and I'll submit a new PR later on.
> 
> **Questions**
> 
> * Would you like me to move the tests in `tests/compas/files/test_stl.py` to `tests/compas/datastructures/test_mesh.py`? Could be fitting to do it in this PR.
> * Would you like me to put the `BaseReader` class in a file with another name, e.g. `_utils.py`?
> * I'm guessing that iterating over lines won't work for the XML reader, am I right in that assumption?

My thoughts on questions : )

- Yep, let's merge the content of test_stl.py into test_from_stl in test_mesh.py.
- It's better to keep as it is for clarity.
- agree, since it's marked-up language.  It would be nice to have an mechanism to throw an error if user tries to load xml in this fashion

**Comment by tetov** [comment]
> * agree, since it's marked-up language.  It would be nice to have an mechanism to throw an error if user tries to load xml in this fashion

There's usually checks in the different modules, e.g. `ply.py` will throw an error if first line is not `'ply'`. It could of course be added to the BaseReader as a general check if you want :).

**Comment by Licini** [comment]
> > * agree, since it's marked-up language.  It would be nice to have an mechanism to throw an error if user tries to load xml in this fashion
> 
> There's usually checks in the different modules, e.g. `ply.py` will throw an error if first line is not `'ply'`. It could of course be added to the BaseReader as a general check if you want :).

That's should be fine then, as long as user be informed the reason of breaking

File: src/compas/files/off.py
Comment (tomvanmele) [inline src/compas/files/off.py]: just out of curiosity, why `RuntimeError`

Comment (tetov) [reply src/compas/files/off.py]: > just out of curiosity, why `RuntimeError`

I reread the builtin exception list and I see now that I doesn't really fit.

TypeError instead?

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py]: Maybe this names should be `def read(mode)` instead of `read_from_location` and `read_binary_from_location`...?

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py]: Instead of reading the whole content and creating an iterator around it, we should do something like:
```python
with self.location.open(mode='r') as fh:
    for line in fh:
        yield line
```

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py]: same comment as above... ;)

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py]: is `64` the block size? perhaps it should be an argument, with a default value.

File: src/compas/files/off.py
Comment (tomvanmele) [reply src/compas/files/off.py]: `TypeError` or `ValueError`

Comment (tomvanmele) [reply src/compas/files/off.py]: i think there are quite a few other occurrences...

Comment (tetov) [reply src/compas/files/off.py]: I know, I might have used the wrong one but at least I was consistently wrong ;).

**Comment by tetov** [comment]
I will change `RuntimeError` to `TypeError` or `ValueError` depending on occurence in all files and move the STL tests to `test_mesh.py` tonight.

I think this should be merged without a new binary reader, and I'll open another PR to start work on that.

File: src/compas/files/off.py
Comment (gonzalocasas) [reply src/compas/files/off.py]: Just `Exception` if you're in doubt. TypeError is not the right one either, because it's not related to the type system.

File: src/compas/files/ply.py
Comment (gonzalocasas) [inline src/compas/files/ply.py]:
```suggestion
        line = next(self.content)

        if line.rstrip() != 'ply':
```

It's better not to mix a conditional with a call that has side effects (such as `next()`) because it might cause bugs when code is refactored at a later time.

**Review by gonzalocasas — CHANGES_REQUESTED** [review]
I like this change! It definitely goes in the right direction of removing code duplication.
Regarding the XML question, it won't really work, we would need a event-based parser like sax or similar, but it's a bit meh....

The main thing I would ask to fix before merging is that now the files are loaded in its entirety and then iterated over lines. This has much higher memory consumption than iterating over the lines directly off of the file object.

You could test this pretty quickly by loading a very large file and printing doing something like, print each line. If the behavior is an iterator, the loading will be almost no time, and then it will start printing one by one; but if the behavior is to load everything up front, you will see a delay before it starts printing.  The alternative is to profile this code when loading large files.

But overall, good stuff!! Keep up the good work!

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py]: Ok!

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py]: Hm, I thought this would do that.. I'll change it!

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py]: Yup, it should. This part is not done and should maybe not be in this PR. I'll merge the functions, add an if check regarding mode and raise NotImplementedError for binary.

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py]: All the file type readers have a read function. Would you still like to use `read` for `BaseReader` and use `read_EXT` (where EXT=file extension)? (See `ply.py` or `obj.py` for good examples)

File: src/compas/files/_BaseReader.py
Comment (tomvanmele) [reply src/compas/files/_BaseReader.py]: no, it converts the list to an iterator, but first makes the list...

File: src/compas/files/_BaseReader.py
Comment (tomvanmele) [reply src/compas/files/_BaseReader.py]: does this work if the file is on a remote location? or does the comment only apply to the `else` part?

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py]: Got it now, sorry for the mistake :).

**Comment by tomvanmele** [comment]
also, perhaps for `dxf` we could build upon `ezdxf`...

**Comment by tetov** [comment]
I'll wait to commit new files when questions raised in @gonzalocasas review and following comments are figured out.

The BaseReader should as pointed out above just yield lines, but I'll also do some profiling to see how much and when the specific modules use up memory.

File: src/compas/files/ply.py
Comment (tetov) [inline src/compas/files/ply.py:429]: This change is because while the definition says `vertex_indices ` Rhino uses `vertex_index ` instead.

**Comment by tetov** [comment]
Per IRL discussion with @brgcode, binary ply has not been implemented, so that test will be commented out.

The code needs to be cleaned up, travis errors fixed but it's slowly getting there.

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py]: @gonzalocasas: what Tom said ☝️

**Comment by tetov** [comment]
Oh and sorry about the amount of commits. Didn’t know a rebase would show up like that on GitHub. I’ll merge next time.

**Comment by tomvanmele** [comment]
@gonzalocasas can we merge this?

**Comment by tetov** [comment]
@gonzalocasas, @brgcode 
https://github.com/compas-dev/compas/pull/348#discussion_r335163659

File: .bumpversion.cfg
Comment (gonzalocasas) [inline .bumpversion.cfg:2]: Why is it show this weird diff? Maybe you need to merge master back into your branch?

File: CHANGELOG.md
Comment (gonzalocasas) [inline CHANGELOG.md:10]: Watch out here, this one also needs to be updated.

File: docs/conf.py
Comment (gonzalocasas) [inline docs/conf.py:22]: same

File: setup.py
Comment (gonzalocasas) [inline setup.py:31]: same

File: src/compas/__init__.py
Comment (gonzalocasas) [inline src/compas/__init__.py:38]: and same here

File: src/compas/__init__.py
Comment (gonzalocasas) [inline src/compas/__init__.py:187]: Why is this changed? I think we're still pull from `raw.githubusercontent.com`...

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py:23]: this should be all lowercase:

```suggestion
        self.is_url = False
```

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py:29]: Don't leave this TODO in the docstring, it will be shown in the pubic docs.

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py:66]: same note here about `TODO` in docstring.

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py:1]: module names should be lower case, and in doing so, you can remove the leading underscore that was probably causing a name collision with your class.

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [inline src/compas/files/_BaseReader.py:81]: This method is great stuff, but I would make a few changes for clarity, because it took me a while to understand what you're doing here:
1. `self.file_signature` is a constant per-class, so I would make that uppercase, eg `self.FILE_SIGNATURE`.
2. Docs! This is a pretty significant addition to the public API, so, it really needs to be well documented.
3. The first check cannot be passed, there are 3 options: 1) the support for urls is implemented 2) not implemented and raises exception, 3) not implemented, but add a `return` immediately.
4. Try to avoid nesting checks. The `else` that contains everything is not needed, because the previous code paths end in error (at least, they should, see previous point)

Then, two questions:
1. Does this work with ascii files (since you open the file in binary mode)?
2. Would it be possible to use `urlretrieve` and actually remove al the remote stuff to be a one-time download into temp and handle everything as files?

File: src/compas/files/amf.py
Comment (gonzalocasas) [inline src/compas/files/amf.py:5]: After the previously suggested rename:
```suggestion
from compas.files.base_reader import BaseReader
```

File: src/compas/files/obj.py
Comment (gonzalocasas) [inline src/compas/files/obj.py:245]: I don't understand this one... is it really correct?

File: src/compas/files/off.py
Comment (gonzalocasas) [inline src/compas/files/off.py:90]: I'm not very convinced about calling these `read_off()`, `read_obj()` and so on. I think it should be the same name across the board. I see that now `read()` is used, so, either rename this level of reads to `read_content` or `read_file`, or rename the `BaseReader.read` to `BaseReader.read_bytes` (or `BaseReader.read_line` or `BaseReader.read_chunk`)

File: src/compas/files/off.py
Comment (gonzalocasas) [inline src/compas/files/off.py:104]: `Import failed` is not very clear, we're not really importing, we're opening files (which will then might be imported into something), so I would be more explicit and limited to this module's functionality in the error message.

File: src/compas/files/ply.py
Comment (gonzalocasas) [inline src/compas/files/ply.py:139]: Is that really a good option to do now? If I understand correctly, binary works before this change, but without numpy so it might be slow; but with this change, it doesn't work anymore. Am I correct?

File: src/compas/files/ply.py
Comment (gonzalocasas) [inline src/compas/files/ply.py:445]: hehe.. is this method really still with a typo on compas Mesh?

File: src/compas/geometry/basic.py
Comment (gonzalocasas) [inline src/compas/geometry/basic.py:63]: This commit actually comes from the rebasing things, right? It's not really part of your branch..?

File: tests/compas/files/test_stl.py
Comment (gonzalocasas) [inline tests/compas/files/test_stl.py:1]: Is this test now done somewhere else?

**Review by gonzalocasas — CHANGES_REQUESTED** [review]
I tried to review this in detail, but the rebase thing gets a lot in the way. There are a few things that are not clear to me, some naming things that should change before merging, the removal of tests, and other remarks; but really, the length of the current PR makes it tricky to review.  If merging master back into this branch does not get rid of all the spurious diffs, then I would suggest you create a new clean branch, and cherry pick your changes on top of a clean status.

**Comment by tetov** [comment]
I’ll open a fresh one (and go through the comments).

File: src/compas/__init__.py
Comment (tetov) [reply src/compas/__init__.py:187]: I’ll have a think about why I did that. I agree but I think I had a reason.

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py:1]: Leading underscore to only make it available inside the module. But maybe not needed with our __init__.py?

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py:81]: Questions:
1. Yes!
2. This goes against what I thought you wanted, to not load a big file into memory?

File: src/compas/files/obj.py
Comment (tetov) [reply src/compas/files/obj.py:245]: Doublecheck.

File: src/compas/files/ply.py
Comment (tetov) [reply src/compas/files/ply.py:139]: Binary ply doesn’t work at all. Double checked with Tom.

Comment (tetov) [reply src/compas/files/ply.py:139]: Would be implemented hopefully in my binary PR.

File: src/compas/geometry/basic.py
Comment (tetov) [reply src/compas/geometry/basic.py:63]: Yup.

File: tests/compas/files/test_stl.py
Comment (tetov) [reply tests/compas/files/test_stl.py:1]: Yes. Moved to test_mesh.py under tests/compas/datastructures.

**Comment by tetov** [comment]
I replied to some. The ones I didn’t reply to I agree to completly. 

I’ll move over your comments to new PR. Thanks for reviewing.

File: tests/compas/files/test_stl.py
Comment (gonzalocasas) [reply tests/compas/files/test_stl.py:1]: ok! I actually think it is more proper to keep this in a `compas/files/test_stl` because it's supposed to be unit test, and by testing the mesh_from_stl, you're actually already integrating a lot of other things; but in this case, it's ok if you want to keep it in the new location, it's not such a big test anyway.. the only thing I would suggest thou is to split the `test_from_stl` method into different ones, each describing what they are actually testing (now it's only a comment but if it fails, you dont see the comment text in the test failure)

File: src/compas/files/ply.py
Comment (gonzalocasas) [reply src/compas/files/ply.py:139]: Ah! alright! yes, now I remember we talked about it at some point (like 3 lifetimes ago!).

File: src/compas/files/_BaseReader.py
Comment (gonzalocasas) [reply src/compas/files/_BaseReader.py:1]: ah ok.. yeah, I think it won't be automatically pulled by the `__init__`, so you could get rid of the leading underscore, but it's also not a big thing if you leave it.

File: src/compas/__init__.py
Comment (gonzalocasas) [reply src/compas/__init__.py:187]: Ah.. I think I see why. If a user reads this and tries to open that URL, he/she gets a 404, while with your change, would get the directory listing, which is what one would expect, so it's good, leave it as you have.

File: src/compas/files/_BaseReader.py
Comment (tetov) [reply src/compas/files/_BaseReader.py:81]: Moved to #418

File: src/compas/files/obj.py
Comment (tetov) [reply src/compas/files/obj.py:245]: Moved to #418.

File: src/compas/files/off.py
Comment (tetov) [reply src/compas/files/off.py:90]: Moved to #418.

File: src/compas/files/off.py
Comment (tetov) [reply src/compas/files/off.py:104]: Moved to #418.

File: src/compas/files/ply.py
Comment (tetov) [reply src/compas/files/ply.py:445]: Hm, we might need to implement mesh.summry().

```
def summry():
    print('No it's rather winntry at the moment')
```

File: tests/compas/files/test_stl.py
Comment (tetov) [reply tests/compas/files/test_stl.py:1]: Moved to #418.

**Review by tetov — COMMENTED** [review]
All unresolved issues moved to #418.

**Pull request closed without merging**
